# Business and data understanding

## Purpose
This notebook contains the business and data understanding according to [2020, Studer et al.](https://arxiv.org/abs/2003.05155) "Towards CRISP-ML(Q): A Machine Learning Process Model with Quality Assurance Methodology".

## Methodology
Besides the methodology described by 2020, Studer et al., I will use the [EDA framework proposed by Tony Ojeda](https://www.youtube.com/watch?v=YEBRkLo568Q).

## WIP - improvements

## Results

## Suggested next steps
- [ ] It was not possible to use the 'cardinalidade' function on 16 attributes. <- Next step: analyze why it happened.


# Setup

## Library import
We import all the required Python libraries

In [2]:
import os

# Data manipulation
import pandas as pd
import numpy as np

# Visualizations
import cufflinks as cf
import matplotlib as plt
import seaborn as sns
import plotly
import plotly.graph_objs as go
import plotly.offline as ply

os.chdir('../')
from src.utils.data_describe import breve_descricao, serie_nulos, cardinalidade
os.chdir('./notebooks/')

# Options for pandas
pd.options.display.max_columns = None
pd.options.display.max_rows = 100

plotly.offline.init_notebook_mode(connected=True)
cf.go_offline(connected=True)
cf.set_config_file(theme='white')

# Autoreload extension
if 'autoreload' not in get_ipython().extension_manager.loaded:
    %load_ext autoreload
    
%autoreload 2

## Parameter definition
We set all relevant parameters for our notebook. By convention, parameters are uppercase, while all the 
other variables follow Python's guidelines.

In [3]:
RAW_FOLDER = '../data/raw/'
RANDOM_STATE = 42


# Data import
We retrieve all the required data for the analysis.

In [4]:
df = pd.read_csv(RAW_FOLDER + 'train.csv')
df.shape

(1460, 81)

## Initial evaluation

In [5]:
# Data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [6]:
serie_nulos(df, corte=0.5)

4 atributos/features/campos possuem mais de 0.5 de valores nulos.


PoolQC         0.995205
MiscFeature    0.963014
Alley          0.937671
Fence          0.807534
dtype: float64

In [7]:
lst_bad_columns = []
lst_good_columns = []

for column in df.select_dtypes(include='object').columns:
    try:
        cardinalidade(df[[column]])
        lst_good_columns.append(column)
    except Exception as e:
        lst_bad_columns.append(column)
        
print(f"""
Using the function 'cardinalidade':
- {len(lst_bad_columns)} columns could not be analyzed;
- {len(lst_good_columns)} columns could be analyzed.
""")


Using the function 'cardinalidade':
- 16 columns could not be analyzed;
- 27 columns could be analyzed.



In [8]:
cardinalidade(df[lst_good_columns])

Atributo  Cardinalidade  \
21     CentralAir              2   
1          Street              2   
4       Utilities              2   
6       LandSlope              3   
24     PavedDrive              3   
16      ExterQual              4   
22    KitchenQual              4   
3     LandContour              4   
2        LotShape              4   
10       BldgType              5   
17      ExterCond              5   
20      HeatingQC              5   
5       LotConfig              5   
0        MSZoning              5   
18     Foundation              6   
19        Heating              6   
12      RoofStyle              6   
26  SaleCondition              6   
23     Functional              7   
9      Condition2              8   
11     HouseStyle              8   
13       RoofMatl              8   
8      Condition1              9   
25       SaleType              9   
14    Exterior1st             15   
15    Exterior2nd             16   
7    Neighborhood             25   

                                              Valores  
21                                             [N, Y]  
1                                        [Grvl, Pave]  
4                                    [AllPub, NoSeWa]  
6                                     [Gtl, Mod, Sev]  
24                                          [N, P, Y]  
16                                   [Ex, Fa, Gd, TA]  
22                                   [Ex, Fa, Gd, TA]  
3                                [Bnk, HLS, Low, Lvl]  
2                                [IR1, IR2, IR3, Reg]  
10              [1Fam, 2fmCon, Duplex, Twnhs, TwnhsE]  
17                               [Ex, Fa, Gd, Po, TA]  
20                               [Ex, Fa, Gd, Po, TA]  
5                 [Corner, CulDSac, FR2, FR3, Inside]  
0                           [C (all), FV, RH, RL, RM]  
18         [BrkTil, CBlock, PConc, Slab, Stone, Wood]  
19              [Floor, GasA, GasW, Grav, OthW, Wall]  
12         [Flat, Gable, Gambrel, Hip, Mansard, Shed]  
26  [Abnorml, AdjLand, Alloca, Family, Normal, Par...  
23            [Maj1, Maj2, Min1, Min2, Mod, Sev, Typ]  
9   [Artery, Feedr, Norm, PosA, PosN, RRAe, RRAn, ...  
11  [1.5Fin, 1.5Unf, 1Story, 2.5Fin, 2.5Unf, 2Stor...  
13  [ClyTile, CompShg, Membran, Metal, Roll, Tar&G...  
8   [Artery, Feedr, Norm, PosA, PosN, RRAe, RRAn, ...  
25  [COD, CWD, Con, ConLD, ConLI, ConLw, New, Oth,...  
14  [AsbShng, AsphShn, BrkComm, BrkFace, CBlock, C...  
15  [AsbShng, AsphShn, Brk Cmn, BrkFace, CBlock, C...  
7   [Blmngtn, Blueste, BrDale, BrkSide, ClearCr, C...

In [9]:
df[lst_bad_columns].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Alley         91 non-null     object
 1   MasVnrType    1452 non-null   object
 2   BsmtQual      1423 non-null   object
 3   BsmtCond      1423 non-null   object
 4   BsmtExposure  1422 non-null   object
 5   BsmtFinType1  1423 non-null   object
 6   BsmtFinType2  1422 non-null   object
 7   Electrical    1459 non-null   object
 8   FireplaceQu   770 non-null    object
 9   GarageType    1379 non-null   object
 10  GarageFinish  1379 non-null   object
 11  GarageQual    1379 non-null   object
 12  GarageCond    1379 non-null   object
 13  PoolQC        7 non-null      object
 14  Fence         281 non-null    object
 15  MiscFeature   54 non-null     object
dtypes: object(16)
memory usage: 182.6+ KB


In [15]:
# Evaluating the int attributes:
cardinalidade(df.select_dtypes(include='int64'))

Atributo  Cardinalidade  \
16   BsmtHalfBath              3   
18       HalfBath              3   
15   BsmtFullBath              4   
22     Fireplaces              4   
17       FullBath              4   
20   KitchenAbvGr              4   
23     GarageCars              5   
33         YrSold              5   
19   BedroomAbvGr              8   
30       PoolArea              8   
4     OverallCond              9   
3     OverallQual             10   
32         MoSold             12   
21   TotRmsAbvGrd             12   
1      MSSubClass             15   
28      3SsnPorch             20   
31        MiscVal             21   
13   LowQualFinSF             24   
6    YearRemodAdd             61   
29    ScreenPorch             76   
5       YearBuilt            112   
27  EnclosedPorch            120   
8      BsmtFinSF2            144   
26    OpenPorchSF            202   
25     WoodDeckSF            274   
12       2ndFlrSF            417   
24     GarageArea            441   
7      BsmtFinSF1            637   
34      SalePrice            663   
10    TotalBsmtSF            721   
11       1stFlrSF            753   
9       BsmtUnfSF            780   
14      GrLivArea            861   
2         LotArea           1073   
0              Id           1460   

                                              Valores  
16                                          [0, 1, 2]  
18                                          [0, 1, 2]  
15                                       [0, 1, 2, 3]  
22                                       [0, 1, 2, 3]  
17                                       [0, 1, 2, 3]  
20                                       [0, 1, 2, 3]  
23                                    [0, 1, 2, 3, 4]  
33                     [2006, 2007, 2008, 2009, 2010]  
19                           [0, 1, 2, 3, 4, 5, 6, 8]  
30             [0, 480, 512, 519, 555, 576, 648, 738]  
4                         [1, 2, 3, 4, 5, 6, 7, 8, 9]  
3                     [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  
32            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]  
21           [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14]  
1   [20, 30, 40, 45, 50, 60, 70, 75, 80, 85, 90, 1...  
28  [0, 23, 96, 130, 140, 144, 153, 162, 168, 180,...  
31  [0, 54, 350, 400, 450, 480, 500, 560, 600, 620...  
13  [0, 53, 80, 120, 144, 156, 205, 232, 234, 360,...  
6   [1950, 1951, 1952, 1953, 1954, 1955, 1956, 195...  
29  [0, 40, 53, 60, 63, 80, 90, 95, 99, 100, 116, ...  
5   [1872, 1875, 1880, 1882, 1885, 1890, 1892, 189...  
27  [0, 19, 20, 24, 30, 32, 34, 36, 37, 39, 40, 42...  
8   [0, 28, 32, 35, 40, 41, 64, 68, 80, 81, 93, 96...  
26  [0, 4, 8, 10, 11, 12, 15, 16, 17, 18, 20, 21, ...  
25  [0, 12, 24, 26, 28, 30, 32, 33, 35, 36, 38, 40...  
12  [0, 110, 167, 192, 208, 213, 220, 224, 240, 25...  
24  [0, 160, 164, 180, 186, 189, 192, 198, 200, 20...  
7   [0, 2, 16, 20, 24, 25, 27, 28, 33, 35, 36, 40,...  
34  [34900, 35311, 37900, 39300, 40000, 52000, 525...  
10  [0, 105, 190, 264, 270, 290, 319, 360, 372, 38...  
11  [334, 372, 438, 480, 483, 495, 520, 525, 526, ...  
9   [0, 14, 15, 23, 26, 29, 30, 32, 35, 36, 38, 39...  
14  [334, 438, 480, 520, 605, 616, 630, 672, 691, ...  
2   [1300, 1477, 1491, 1526, 1533, 1596, 1680, 186...  
0   [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...

In [48]:
lst_time = [x for x in df.columns if ('yr' in x.lower()) or ('year' in x.lower())]
# After reading the data description, I realized that 'MoSold' is a time attribute too.
lst_time.append('MoSold')

print(f"""There is/are {len(lst_time)} time attributes:
{lst_time}""")

There is/are 5 time attributes:
['YearBuilt', 'YearRemodAdd', 'GarageYrBlt', 'YrSold', 'MoSold']


In [46]:
lst_area = [x for x in df.columns if ('area' in x.lower()) or ('sf' in x.lower())]

print(f"""There is/are {len(lst_area)} area attributes:
{lst_area}""")

There are 14 area attributes:
['LotArea', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'PoolArea']


In [47]:
lst_float = [
    x for x in df.select_dtypes(include='float64').columns.tolist() if (x not in lst_area) and (x not in lst_time)
]

print(f"""There is/are {len(lst_float)} float attributes:
{lst_float}""")

There is/are 1 float attributes:
['LotFrontage']


### Partial conclusions:
- From the 91 attributes, we have:
 - float64(3), int64(35), object(43)

- There are 4 attributes with more than 50% of null values:
 - PoolQC         0.995205
 - MiscFeature    0.963014
 - Alley          0.937671
 - Fence          0.807534
 
- It was not possible to use the 'cardinalidade' function on 16 attributes. <- Next step.

**Action:**

**30/05/2022**:
- The field 'id' will be dropped.
- The 4 attributes wit more than 80% of null values will be dropped.
- For the baseline model, I will take only the numerical, and time fields (
 - 5 time attributes;
 - 14 area attributes; and
 - 1 float attributes.

# EDA framework

<img src="../references/eda_framework.png" alt="eda" class="bg-primary" width="500px">

In [10]:
df.sample(10, random_state=RANDOM_STATE)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
892    893          20       RL         70.0     8414   Pave   NaN      Reg   
1105  1106          60       RL         98.0    12256   Pave   NaN      IR1   
413    414          30       RM         56.0     8960   Pave  Grvl      Reg   
522    523          50       RM         50.0     5000   Pave   NaN      Reg   
1036  1037          20       RL         89.0    12898   Pave   NaN      IR1   
614    615         180       RM         21.0     1491   Pave   NaN      Reg   
218    219          50       RL          NaN    15660   Pave   NaN      IR1   
1160  1161         160       RL         24.0     2280   Pave   NaN      Reg   
649    650         180       RM         21.0     1936   Pave   NaN      Reg   
887    888          50       RL         59.0    16466   Pave   NaN      IR1   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
892          Lvl    AllPub    Inside       Gtl       Sawyer       Norm   
1105         Lvl    AllPub    Corner       Gtl      NoRidge       Norm   
413          Lvl    AllPub    Inside       Gtl      OldTown     Artery   
522          Lvl    AllPub    Corner       Gtl      BrkSide      Feedr   
1036         HLS    AllPub    Inside       Gtl       Timber       Norm   
614          Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
218          Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
1160         Lvl    AllPub    Inside       Gtl      NPkVill       Norm   
649          Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
887          Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
892        Norm     1Fam     1Story            6            8       1963   
1105       Norm     1Fam     2Story            8            5       1994   
413        Norm     1Fam     1Story            5            6       1927   
522        Norm     1Fam     1.5Fin            6            7       1947   
1036       Norm     1Fam     1Story            9            5       2007   
614        Norm   TwnhsE     SFoyer            4            6       1972   
218        Norm     1Fam     1.5Fin            7            9       1939   
1160       Norm    Twnhs     2Story            6            5       1978   
649        Norm    Twnhs     SFoyer            4            6       1970   
887        Norm     1Fam     1.5Fin            5            7       1955   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
892           2003       Hip  CompShg     HdBoard     HdBoard       None   
1105          1995     Gable  CompShg     HdBoard     HdBoard    BrkFace   
413           1950     Gable  CompShg     WdShing     Wd Shng       None   
522           1950     Gable  CompShg     CemntBd     CmentBd       None   
1036          2008       Hip  CompShg     VinylSd     VinylSd      Stone   
614           1972     Gable  CompShg     CemntBd     CmentBd       None   
218           2006     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1160          1978     Gable  CompShg     Plywood     Brk Cmn       None   
649           1970     Gable  CompShg     CemntBd     CmentBd       None   
887           1955     Gable  CompShg     MetalSd     MetalSd       None   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
892          0.0        TA        TA     CBlock       TA       TA   
1105       362.0        Gd        TA      PConc       Ex       TA   
413          0.0        TA        TA     CBlock       TA       TA   
522          0.0        TA        Gd     CBlock       TA       TA   
1036        70.0        Gd        TA      PConc       Ex       TA   
614          0.0        TA        TA     CBlock       Gd       TA   
218        312.0        Gd        Gd     CBlock       TA       TA   
1160         0.0        TA        TA     CBlock       Gd       TA   
649          0.0        TA  

## Identity

### Types of information

### Entities in dataset

## Review

### Transformation methods

### Visualization methods

## Create

### Category aggregations

### Continuous bins

### Cluster categories

# Data processing
Put here the core of the notebook. Feel free di further split this section into subsections.

# References
We report here relevant references:
1. author1, article1, journal1, year1, url1
2. author2, article2, journal2, year2, url2